Expanded discussion of Provenance, rearranged sections (#12)
confidential-computing/twi-wimse: draft-twi-reference-architecture.md

      org: Confidential Computing Consortium Trustworthy Workload Identity SIG

--- abstract

> This Internet-Draft covers a gap analysis performed by the Confidential Computing Consortium on WIMSE identifying areas where the current WIMSE Architecture should be extended to accommodate Workloads running in Confidential Computing environments. It outlines high-level requirements for the these extensions and describes a series of use cases.

Reviewer: Capitalizing terms that are not introduced yet in the abstract is not a huge problem, but is typically started after the terminology section. I am not objecting. At this point the "issue is basically "immaterial".

Author: I am going to opt for consistency of capitalization of certain words throughout the document. We can always change that later.

Current ways of issuing and managing workload identities, as well as those required for effective protection of data-in-use, suffer from several architectural shortcomings; chief among them:

1. Lack of Workload isolation against the hardware and the operating system owners/administrators, as well as peer Workload instances
2. Lack of strong binding between a Workload Credential and the Workload instance to which that Credential had been issued
3. Inability to associate a Credential with a set of decisions leading up to its issuance
_Note that these requirements are related: lack of process isolation eases credential exfiltration and leads to credential leakage and reuse._

> The Confidential Computing Consortium defines Confidential Computing as "protection of data in use by performing computation in a hardware-based, attested Trusted Execution Environment (TEE)". In the immediate term, effective Confidential Computing faces a significant challenge: the scale of uplift and immature tooling are both clear barriers. Longer term, however, Confidential Computing provides a vital improvement due to its unique features and broad availability. Confidential Computing-assisted mechanisms have to fit inside the emerging Workload Identity solution ecosystem. The evolution of this ecosystem should remain in alignment with the expectations of the owners and operators of Confidential Computing Workloads. These efforts will build on the concept of _Trustworthy Workload Identity_ defined further below. Data-in-use protection of Workloads that have such Identities will be a critical downstream effect.

Author: > scale of uplift and immature tooling might require a representative reference at this early point of the document

I have removed that wording as it is not needed for the draft

> The Confidential Computing Consortium defines Confidential Computing as "protection of data in use by performing computation in a hardware-based, attested Trusted Execution Environment (TEE)". In the immediate term, effective Confidential Computing faces a significant challenge: the scale of uplift and immature tooling are both clear barriers. Longer term, however, Confidential Computing provides a vital improvement due to its unique features and broad availability. Confidential Computing-assisted mechanisms have to fit inside the emerging Workload Identity solution ecosystem. The evolution of this ecosystem should remain in alignment with the expectations of the owners and operators of Confidential Computing Workloads. These efforts will build on the concept of _Trustworthy Workload Identity_ defined further below. Data-in-use protection of Workloads that have such Identities will be a critical downstream effect.

Author: > Using the named word "identities" is often avoided (because it attracts trouble sometimes. I assume that referencing wimse reduces that issue, also including our scoped definition in the TWI I-D helps.

What are you proposing we change in response to this comment? I'm not clear.

> While it is quite clear that for the foreseeable future this need is not going to be met by Confidential Computing alone, the security features related to Workload identification and isolation offered by Confidential Computing eclipse other approaches. Confidential Computing-strengthened Workload Identity will need to be designed to fit within, and ideally serve as the North Star of Workload Identity as it becomes a foundational pillar of trustworthy and governable enterprise computing.

Reviewer: use of "this". Repetition of the problem statement is fine.

Author: > use of "this". Repetition of the problem statement is fine.

What do you mean by 'use of "this"'?

# Gap Analysis

> An analysis was performed by the Confidential Computing Consortium of the existing WIMSE Architecture to identify extensions necessary to meet the level of trustworthiness required by Confidential Computing environments.

Author: I don't believe the CCC defines "trustworthiness levels".

> An analysis was performed by the Confidential Computing Consortium of the existing WIMSE Architecture to identify extensions necessary to meet the level of trustworthiness required by Confidential Computing environments.

Author: > 

The proposed wording does not clarify things (for me at least) and contains a couple of typos. I kept some of it in my reworded suggestion.

      org: Confidential Computing Consortium Trustworthy Workload Identity SIG

--- abstract

> This Internet-Draft covers a gap analysis performed by the Confidential Computing Consortium on WIMSE identifying areas where the current WIMSE Architecture should be extended to accommodate Workloads running in Confidential Computing environments. It outlines high-level requirements for the these extensions and describes a series of use cases.

Reviewer 2: ```suggestion
This Internet-Draft covers a gap analysis performed by the Confidential Computing Consortium on WIMSE identifying areas where the current WIMSE Architecture should be extended to accommodate Workloads running in Confidential Computing environments. It outlines high-level requirements for these extensions and describes a series of use cases.
```

Current ways of issuing and managing workload identities, as well as those required for effective protection of data-in-use, suffer from several architectural shortcomings; chief among them:

1. Lack of Workload isolation against the hardware and the operating system owners/administrators, as well as peer Workload instances
2. Lack of strong binding between a Workload Credential and the Workload instance to which that Credential had been issued
3. Inability to associate a Credential with a set of decisions leading up to its issuance

_Note that these requirements are related: lack of process isolation eases credential exfiltration and leads to credential leakage and reuse._

> The Confidential Computing Consortium defines Confidential Computing as "protection of data in use by performing computation in a hardware-based, attested Trusted Execution Environment (TEE)". Confidential Computing can address the existing shortcomings in a way that is compatible with the emerging Workload Identity solution ecosystem and can evolve in alignment with the expectations of the owners and operators of Confidential Computing Workloads. These efforts will build on the concept of _Trustworthy Workload Identity_ defined further below. Data-in-use protection of Workloads that have such Identities will be a critical downstream effect.

Reviewer 2: ```suggestion
The Confidential Computing Consortium defines Confidential Computing as "protection of data in use by performing computation in a hardware-based, attested Trusted Execution Environment (TEE)". Confidential Computing can address the existing shortcomings in a way that is compatible with the emerging Workload Identity solution ecosystem and can evolve in alignment with the expectations of the owners and operators of Confidential Computing Workloads. These efforts will build on the concept of _Trustworthy Workload Identity_ (TWI) defined further below. Data-in-use protection of Workloads that have such Identities will be a critical downstream effect.
```

Reviewer 2: I just defined the acronym at the first use

* **Workload Provenance** (((Workload Provenance))) is a linkage between a Workload Credential and a trusted entity (e.g., a vendor, developer, or issuer) responsible for the creation and/or attestation of the corresponding Workload.

In the light of the definitions just given, a Workload Identity is said to be Trustworthy iff the following three properties hold true:

1. The Workload is confidentiality- and integrity-isolated from its hosting environment as well as peer workloads throughout the lifetime of the Workload instance
2. Any credential representing a Workload Identity is bound to that Workload instance
3. A Workload Credential can be traced back to the Workload’s Provenance and the credential issuer

# Integration of Confidential Computing into WIMSE
**Secure Key Storage & Cryptographic Operations:**

> With TEE’s, a Workload’s private keys and sensitive cryptographic operations (such as signing or validating tokens) can be isolated from the hosting environment, reducing the risk of key leakage even if the surrounding system is compromised. (For instance, the key backing a WIMSE token — be it a WIT or an X.509 certificate — can be generated within a TEE, ensuring that the proof-of-possession mechanism remains intact.) Alternatively, a key for a WIT or an X.509 certificate could be obtained from an HSM contingent on successful remote attestation of a Workload instance.

Reviewer 3: The `HSM` piece is a little fuzzy, esp. as I'm not aware of any HSM offering capable of runtime attestation of a TEE/enclave (which would be required for it to issue any kind of key).

On the other hand, if the attestation was to be unidirectional only (TWI attests 'its' crypto provider before asking for key), then:
  - If the HSM is 'just' a stateless key **generator**, it feels like we can just wrap it under "TEE generates the key <somehow> umbrella", w/o over defining.
  - If the HSM has state (and this is a key **sharing** concept), we have introduced new attack surface (non-CC, it suffices to break the HSM auth token).

* **Workload Provenance** (((Workload Provenance))) is a linkage between a Workload Credential and a trusted entity (e.g., a vendor, developer, or issuer) responsible for the creation and/or attestation of the corresponding Workload.

In the light of the definitions just given, a Workload Identity is said to be Trustworthy iff the following three properties hold true:

1. The Workload is confidentiality- and integrity-isolated from its hosting environment as well as peer workloads throughout the lifetime of the Workload instance
2. Any credential representing a Workload Identity is strongly bound to that Workload instance

> 3. A Workload Credential can be traced back to the Workload’s Provenance and the credential issuer

Author: ```suggestion
3. A Workload Credential can be traced back to the Workload’s Provenance, either at runtime or for a later audit
```

**Protected Credential Exchange:**
For the Credential exchange patterns defined in the WIMSE Credential Exchange draft, Confidential Computing can provide a secure TEE in which the exchange logic runs. This ensures that the process of exchanging or re-provisioning credentials is protected against tampering and eavesdropping.

**Mitigating Runtime Compromise:**
Incorporating confidential computing within the workload’s execution environment can lower the risk that runtime attacks (such as memory scraping or side-channel attacks) can expose critical identity or authentication tokens. For example, the confidential computing environment can be used to securely generate and verify proofs of possession that are important within the WIMSE authentication protocol.

**Cross-domain Trust Enhancement:**
In multi-cloud and cross-trust-domain interactions, a hardware-rooted attestation from a TEE can serve as an independent trust anchor. This extra layer of verification helps ensure that even if traditional software-based checks fail, the actual Workload Identity remains trustworthy.

> # TWI Use Cases

Reviewer 2: Here I was seeking Practical Real World Use Cases: That Demonstrates the Value of this Work Item.

## Provenance

### Workload Provenance
Workload Provenance is the metadata pertaining to the Workload. Provenance MAY contain information about a variety of aspects of inputs and decisions that went into the creation of a Workload, as below.
1. Workload Composition, i.e. a detailed list of components that comprise a Workload. This could be expressed as a Software Bill of Materials expressed in terms of industry standards, like SPDX or CycloneDX <TODO: add references?>.

2. Details about the Continous Integration (CI) or Continous Delivery (CD) of Workload.

3. Set of compliance tests that executed on the Workload.

4. Details of Vendor/SaaS information.

The policy for assessing or auditing Credentials MAY demand the information about the Provenance of the Workload. This requries work in three areas:

>    1. Collection: the process of compiling Provenance information about the Workload during the processes that lead up to the creation of a Credential

Reviewer: Maybe consider:
1. Generation, which includes Collection
2. Conveyance
3. Appraisal, which fuels Utilization

Reviewer 3: +1
I'd also suggest decoupling 'workload creation' from 'credential creation'
